### PR TITLE
DOC Use FieldValidators for FormFieldValidation

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11422

### CODE REVIEW

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 1]
```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
 
 > [!NOTE]
 > Each individual [FormField](api:SilverStripe\Forms\FormField) instance is responsible for validating the submitted content through the
-> [FormField::validate()](api:SilverStripe\Forms\FormField::validate()) method. By default, this just checks the value exists. Fields like `EmailField` override
-> `validate` to check for a specific format.
+> [FormField::validate()](api:SilverStripe\Forms\FormField::validate()) method. Fields like [`EmailField::validate()`](api:SilverStripe\Forms\EmailField::validate()) check for a specific format.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]: `EmailField::validate()` is not a field - either change back to the wording used before, or update reflect that you're not talking about the field, but the method _on_ the field.

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 2]
```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
...
 class FormFieldValidationExtension extends Extension
 {
-    protected function updateValidationResult(bool &$result, Validator $validator)
+    protected function updateValidate(ValidationResult $result): void
     {
         if (str_ends_with($this->owner->Value(), '@example.com')) {
-            $validator->validationError($this->owner->Name(), 'Please provide a valid email address');
-            $result = false;
+            $result->addFieldError(
+                $this->owner->Name(),
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
                $this->getOwner()->Name(),
```
Best practice is to use public getters over protected properties

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 3]
```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
...
+    public function validate(): ValidationResult
     {
-        if ((int) $this->Value() === 10) {
-            $validator->validationError($this->Name(), 'This value cannot be 10');
-            return $this->extendValidationResult(false, $validator);
-        }
-
-        return $this->extendValidationResult(true, $validator);
+        $result = ValidationResult::create();
+        $this->beforeExtending('updateValidate', function () use ($result) {
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
        $this->beforeExtending('updateValidate', function (ValidationResult $result) {
```

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 4]
```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
...
+            if ((int) $this->Value() === 10) {
+                $result->addFieldError($this->Name(), 'This value cannot be 10');
+            }
+        });
+        return $result->combineAnd(parent::validate());
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
        return parent::validate();
```

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 5]
```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
...
     }
-}
 ```
 
-The `validate` method should compute a boolean (`true` if the value passes validation and `false` if Silverstripe CMS
-should trigger a validation error on the page) and pass this to the `extendValidationResult` method to allow extensions
-to hook into the validation logic. In addition, in the event of failed validation, a useful error message must be set
-on the given validator.
+The `validate()` method must create a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object and add any errors to it, which should contain a useful error message. The `validate()` method should utilise the `$this->beforeExtending('updateValidate', ...` block and be followed by `return $result->combineAnd(parent::validate());` in order for code in the parent class to be executed in the correct order.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]: It doesn't have to create one, but it must return one.
Best practice if we're using `beforeExtending()` is to use the one passed into the extension hook, which the parent class creates.

File: en/08_Changelogs/6.0.0.md [hunk 6]
```diff
@@ -790,6 +790,14 @@ PHP code such as resolvers that were in [`silverstripe/asset-admin`](http://gith
 
 If your project does not have any custom GraphQL, after upgrading you may still have the old `.graphql-generated` and `public/_graphql` folders in your project. You can safely remove these folders.
 
+### `FormField` classes now use `FieldValidator` for validation
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md]: Add this to the links list at the top of the page, and give it a short anchor

File: en/08_Changelogs/6.0.0.md [hunk 7]
```diff
@@ -790,6 +790,14 @@ PHP code such as resolvers that were in [`silverstripe/asset-admin`](http://gith
...
+
+Many of [`FormField`](api:SilverStripe\Forms\FormField) subclasses in the `SilverStripe\Forms` namespace now use `FieldValidator` classes for validation, which are also used for `DBField` validation. This has meant that much of the old `validate()` logic on `FormField` subclasses has been removed as it was duplicated in the `FieldValidator` classes. Some custom in `validate()` methods not found in `FieldValidator` classes methods has been retained.
+
+As part of this change, the [`FormField::validate()`](api:SilverStripe\Forms\FormField::validate()) now returns a `ValidationResult` object rather than a `bool` which it previously did. Also, the `$validator` parameter has been removed. If you have implemented a custom `validate()` method in a `FormField` subclass, you will need to update it to return a `ValidationResult` object instead and remove the `$validator` parameter.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md]:
```suggestion
As part of this change, the [`FormField::validate()`](api:SilverStripe\Forms\FormField::validate()) now returns a `ValidationResult` object where it used to return a boolean. The `$validator` parameter has also been removed. If you have implemented a custom `validate()` method in a `FormField` subclass, you will need to update it to return a `ValidationResult` object instead and remove the `$validator` parameter.
```

File: en/08_Changelogs/6.0.0.md [hunk 8]
```diff
@@ -790,6 +790,14 @@ PHP code such as resolvers that were in [`silverstripe/asset-admin`](http://gith
...
+
+Also, the `extendValidationResult()` method and the `updateValidationResult` extension hook on `FormField` have both been removed and replaced with an `updateValidate` hook instead which has a single `ValidationResult $result` parameter. This matches the `updateValidate` extension hook on `DataObject`.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md]:
```suggestion
The `extendValidationResult()` method and the `updateValidationResult` extension hook on `FormField` have both been removed and replaced with an `updateValidate` hook instead, which has a single `ValidationResult $result` parameter. This matches the `updateValidate` extension hook on `DataObject`.
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 2 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 3 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 4 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 5 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/08_Changelogs/6.0.0.md (hunk 6 above)
Comment (emteknetnz) [reply en/08_Changelogs/6.0.0.md]: Updated

File: en/08_Changelogs/6.0.0.md (hunk 7 above)
Comment (emteknetnz) [reply en/08_Changelogs/6.0.0.md]: Updated

File: en/08_Changelogs/6.0.0.md (hunk 8 above)
Comment (emteknetnz) [reply en/08_Changelogs/6.0.0.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 1 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Was updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 9]
```diff
@@ -71,71 +71,93 @@ In this example we will be required to input a value for `Name` and a valid emai
...
+        return parent::validate();
     }
-}
 ```
 
-The `validate` method should compute a boolean (`true` if the value passes validation and `false` if Silverstripe CMS
-should trigger a validation error on the page) and pass this to the `extendValidationResult` method to allow extensions
-to hook into the validation logic. In addition, in the event of failed validation, a useful error message must be set
-on the given validator.
+The `validate()` method returns a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object with any errors added to it, and should contain a useful error message. The `validate()` method should follow the example aboe and utilise the `$this->beforeExtending('updateValidate', ...` block and be followed by `return parent::validate();` in order for code in the parent class to be executed in the correct order.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
The `validate()` method returns a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object with any errors added to it, and should contain a useful error message. The `validate()` method should follow the example above and utilise the `$this->beforeExtending('updateValidate', ...` block and be followed by `return parent::validate();` in order for code in the parent class to be executed in the correct order.
```

Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
The `validate()` method returns a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object with any errors added to it, and should contain a useful error message. The `validate()` method should follow the example aboe and utilise the `$this->beforeExtending('updateValidate', ...)` method and be followed by `return parent::validate();` in order for code in the parent class to be executed in the correct order.
```

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 10]
```diff
@@ -71,71 +71,93 @@ In this example we will be required to input a value for `Name` and a valid emai
 
 > [!NOTE]
 > Each individual [FormField](api:SilverStripe\Forms\FormField) instance is responsible for validating the submitted content through the
-> [FormField::validate()](api:SilverStripe\Forms\FormField::validate()) method. By default, this just checks the value exists. Fields like `EmailField` override
-> `validate` to check for a specific format.
+> [FormField::validate()](api:SilverStripe\Forms\FormField::validate()) method. For example [`EmailField::validate()`](api:SilverStripe\Forms\EmailField::validate()) check the field value is a valid email.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
> [FormField::validate()](api:SilverStripe\Forms\FormField::validate()) method. For example [`EmailField::validate()`](api:SilverStripe\Forms\EmailField::validate()) checks the field value is a valid email.
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Just some minor changes

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 9 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 10 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

File: en/02_Developer_Guides/03_Forms/01_Validation.md [hunk 11]
```diff
@@ -71,71 +71,95 @@ In this example we will be required to input a value for `Name` and a valid emai
...
+        return parent::validate();
     }
 }
 ```
 
-The `validate` method should compute a boolean (`true` if the value passes validation and `false` if Silverstripe CMS
-should trigger a validation error on the page) and pass this to the `extendValidationResult` method to allow extensions
-to hook into the validation logic. In addition, in the event of failed validation, a useful error message must be set
-on the given validator.
+The `validate()` method returns a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object with any errors added to it, and should contain a useful error message. The `validate()` method should follow the example above and utilise the `$this->beforeExtending('updateValidate', ...` method and be followed by `return parent::validate();` in order for code in the parent class to be executed in the correct order.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/03_Forms/01_Validation.md]:
```suggestion
The `validate()` method returns a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object with any errors added to it, and should contain a useful error message. The `validate()` method should follow the example above and utilise the `$this->beforeExtending('updateValidate', ...)` method and be followed by `return parent::validate();` in order for code in the parent class to be executed in the correct order.
```
As per https://github.com/silverstripe/developer-docs/pull/621#discussion_r1855636480

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: en/02_Developer_Guides/03_Forms/01_Validation.md (hunk 11 above)
Comment (emteknetnz) [reply en/02_Developer_Guides/03_Forms/01_Validation.md]: Updated

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge when all code PRs are merged

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**